archive/100-theorems-list/81_sum_of_prime_reciprocals_diverges.lean

```diff
@@ -0,0 +1,221 @@
+
+lemma not_tendsto_imp_sum_lt_half
+  (h : ¬ tendsto (λ n, ∑ p in {p ∈ range n | nat.prime p}, (1 / (p : ℝ))) at_top at_top) :
+  ∃ k, ∀ x, ∑ p in {p ∈ range (x + 1) | k < p ∧ nat.prime p}, 1 / (p : ℝ) < 1 / 2 :=
+begin
+  have h0 : (λ n, ∑ p in {p ∈ range n | nat.prime p}, (1 / (p : ℝ)))
+          = λ n, ∑ p in range n, ite (nat.prime p) (1 / (p : ℝ)) 0,
+  { simp only [sum_filter, filter_congr_decidable, sep_def] },
+
+  have hf : ∀ n : ℕ, 0 ≤ ite (nat.prime n) (1 / (n : ℝ)) 0,
+  { intro n, split_ifs,
+    { simp only [one_div, inv_nonneg, nat.cast_nonneg] },
+    { exact rfl.ge } },
+
+  rw [h0, ← summable_iff_not_tendsto_nat_at_top_of_nonneg hf, summable_iff_vanishing] at h,
+  obtain ⟨s, h⟩ := h (set.Ioo (-1) (1/2)) (mem_nhds_sets is_open_Ioo (by norm_num)),
+  obtain ⟨k, hk⟩ := exists_nat_subset_range s,
+  use k,
+  intro x,
+
+  let P := {p ∈ range (x + 1) | k < p ∧ nat.prime p},
+  have hP : P = filter (λ p : ℕ, k < p ∧ nat.prime p) (range (x + 1)),
+  { simp only [P, sep_def, filter_congr_decidable] },
+
+  specialize h (filter (λ n : ℕ, k < n) (range (x + 1))) _,
+  { rw disjoint_iff_ne,
+    simp_intros a ha b hb only [mem_filter],
+    exact ((mem_range.mp (hk hb)).trans ha.2).ne' },
+  { simpa only [←sum_filter, filter_filter, ←hP] using h.2 },
+end
+
+lemma range_x_sdiff_M_eq_U {x k : ℕ} :
+  range x \ {e ∈ range x | ∀ p : ℕ, (nat.prime p ∧ p ∣ e + 1) → p ≤ k} =
+  finset.bUnion {p ∈ range (x + 1) | k < p ∧ nat.prime p} (λ p, {e ∈ range x | p ∣ e + 1}) :=
+begin
+  ext e,
+  simp only [mem_bUnion, not_and, mem_sdiff, sep_def, mem_filter, mem_range],
+  push_neg,
+  split,
+  { rintros ⟨hex, hexh⟩,
+    obtain ⟨p, ⟨hpp, hpe1⟩, hpk⟩ := hexh hex,
+    refine ⟨p, _, ⟨hex, hpe1⟩⟩,
+    simp only [mem_filter, mem_range],
+    exact ⟨(nat.le_of_dvd e.succ_pos hpe1).trans_lt (nat.succ_lt_succ hex), hpk, hpp⟩ },
+  { rintros ⟨p, hpfilter, ⟨hex, hpe1⟩⟩,
+    simp only [mem_filter, mem_range] at hpfilter,
+    rw imp_iff_right hex,
+    exact ⟨hex, ⟨p, ⟨hpfilter.2.2, hpe1⟩, hpfilter.2.1⟩⟩ },
+end
+
+lemma card_U_le_x_mul_sum {x k : ℕ} :
+  (card
+    (finset.bUnion {p ∈ range (x + 1) | k < p ∧ nat.prime p} (λ p, {e ∈ range x | p ∣ e + 1}))
+  : ℝ)
+  ≤ x * ∑ p in {p ∈ range (x + 1) | k < p ∧ nat.prime p}, 1 / p :=
+begin
+  let P := {p ∈ range (x + 1) | k < p ∧ nat.prime p},
+  let N := λ p, {e ∈ range x | p ∣ e + 1},
+  have h : card (finset.bUnion P N) ≤ ∑ p in P, card (N p) := card_bUnion_le,
+
+  calc  (card (finset.bUnion P N) : ℝ)
+      ≤ ∑ p in P, card (N p)  : by assumption_mod_cast
+  ... ≤ ∑ p in P, x / p       : by { refine sum_le_sum (λ p, _),
+                                     simp only [P, sep_def, mem_filter, mem_range],
+                                     rintros ⟨-, -, hpp⟩,
+                                     simp only [N, card_le_div_real (nat.prime.pos hpp)] }
+  ... = ∑ p in P, x * (1 / p) : finset.sum_congr rfl (λ p, λ _, div_eq_mul_one_div x p)
```

Comment:
You're making this overly complicated  - `mul_one_div` is your friend.
Something like this should work (no new step in the `calc` bloc needed):
```suggestion
                                     simp only [P, sep_def, mem_filter, mem_range, N, mul_one_div],
                                     rintros ⟨-, -, hpp⟩,
                                     exact card_le_div_real (nat.prime.pos hpp) }
```